3party/opening_hours/opening_hours.hpp

```diff
@@ -0,0 +1,563 @@
+
+  Copyright (c) 2015 Mail.Ru Group
+
+  Permission is hereby granted, free of charge, to any person obtaining a copy
+  of this software and associated documentation files (the "Software"), to deal
+  in the Software without restriction, including without limitation the rights
+  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+  copies of the Software, and to permit persons to whom the Software is
+  furnished to do so, subject to the following conditions:
+
+  The above copyright notice and this permission notice shall be included in
+  all copies or substantial portions of the Software.
+
+  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+  THE SOFTWARE.
+*/
+
+#pragma once
+
+#include <chrono>
+#include <iostream>
+#include <string>
+#include <type_traits>
+#include <vector>
+
+namespace osmoh
+{
+
+class Time
+{
+  enum State
+  {
+    STATE_IS_NOT_TIME = 0,
+    STATE_HAVE_HOURS = 1,
+    STATE_HAVE_MINUTES = 2,
+  };
+
+  using TStateRep = std::underlying_type<State>::type;
+
+public:
+  enum class Event
+  {
+    NotEvent,
+    Dawn,
+    Sunrise,
+    Sunset,
+    Dusk
+  };
+
+  using THours = std::chrono::hours;
+  using TMinutes = std::chrono::minutes;
+
+  Time() = default;
+  Time(Time const &) = default;
+  Time(THours const hours);
+  Time(TMinutes const minutes);
+
+  Time & operator=(Time const &) = default;
+
+  THours::rep GetHoursCount() const;
+  TMinutes::rep GetMinutesCount() const;
+
+  THours GetHours() const;
+  TMinutes GetMinutes() const;
+
+  void SetHours(THours const hours);
+  void SetMinutes(TMinutes const minutes);
+
+  Event GetEvent() const {return m_event;}
+  void SetEvent(Event const event);
+
+  bool IsEvent() const;
+  bool IsEventOffset() const;
+  bool IsHoursMinutes() const;
+  bool IsMinutes() const;
+  bool IsTime() const;
+  bool HasValue() const;
+
+  Time operator+(Time const & t);
+  Time operator-(Time const & t);
+  Time & operator-();
+
+private:
+  Time GetEventTime() const;
+
+  Event m_event = Event::NotEvent;
+  TMinutes m_duration = TMinutes::zero();
+  TStateRep m_state = STATE_IS_NOT_TIME;
+};
+
+inline constexpr Time::THours operator ""_h(unsigned long long int h)
+{
+  return Time::THours(h);
+}
+
+inline constexpr Time::TMinutes operator ""_min(unsigned long long int m)
+{
+  return Time::TMinutes(m);
+}
+
+std::ostream & operator<<(std::ostream & ost, Time::Event const event);
+std::ostream & operator<<(std::ostream & ost, Time const & time);
+
+class Timespan
+{
+public:
+  Timespan() = default;
+  Timespan(Timespan const &) = default;
```

Comment:
If this class will not have external operators, all single-argument constructors should be marked as 'explicit'.